### PR TITLE
WriteThread::EnterAsBatchGroupLeader reorder writers

### PR DESCRIPTION
**Pull request by mm304321141**
Fix issue #126

### CODE REVIEW

**Review by oujiaxin — APPROVED** [review]

**Comment by mm304321141** [comment]
@siying @riversand963 @ajkr @skyzh @v1siuol 
We refactor `WriteThread::EnterAsBatchGroupLeader` for **strict** `WriteOptions::disableWAL`
pls give some suggestions

File: db/write_thread.cc [hunk 1]
```diff
@@ -428,51 +428,79 @@ size_t WriteThread::EnterAsBatchGroupLeader(Writer* leader,
...
-
-    auto batch_size = WriteBatchInternal::ByteSize(w->batch);
-    if (size + batch_size > max_size) {
-      // Do not make batch too big
-      break;
+  }
+  // append r_list after write_group end
+  if (rb != nullptr) {
+    rb->link_older = we;
+    re->link_newer = nullptr;
```
Comment (skyzh) [inline db/write_thread.cc:494]: Seems redundant. Shouldn't `re->link_newer` be `nullptr` by default?

**Comment by yapple** [comment]
> @siying @riversand963 @ajkr @skyzh @v1siuol
> We refactor `WriteThread::EnterAsBatchGroupLeader` for **strict** `WriteOptions::disableWAL`
> pls give some suggestions

File: db/write_thread.cc [hunk 2]
```diff
@@ -428,51 +428,79 @@ size_t WriteThread::EnterAsBatchGroupLeader(Writer* leader,
...
+    we->link_newer = rb;
+    if (!newest_writer_.compare_exchange_weak(w, re)) {
+      while (w->link_older != newest_writer) {
```
Comment (skyzh) [inline db/write_thread.cc:497]: Can we use `newest_writer->link_newer->link_order = re` to avoid this while loop?

Comment (mm304321141) [reply db/write_thread.cc:497]: no, `newest_writer->link_newer` will be nullptr, seek code of `LinkOne` & `CreateMissingNewerLinks`

File: db/write_thread.cc (hunk 1 above)
Comment (mm304321141) [reply db/write_thread.cc:494]: `re` may not newest_writer ... it can be removed from middle node, and we haven't reset it to null when we remove it

**Pull request merged**